src/lib/Cashlink.ts

```diff
@@ -0,0 +1,453 @@
+    contactName?: string; /** unused for now */
+}
+
+export default class Cashlink {
+    get value() {
+        return this._value || 0;
+    }
+
+    set value(value: number) {
+        if (this._value && (this._immutable || this.state !== CashlinkState.UNCHARGED)) {
+            throw new Error('Cashlink is immutable');
+        }
+        if (!Nimiq.NumberUtils.isUint64(value) || value === 0) throw new Error('Malformed value');
+        this._value = value;
+    }
+
+    get message() {
+        return Utf8Tools.utf8ByteArrayToString(this._messageBytes);
+    }
+
+    set message(message) {
+        if (this._messageBytes.byteLength && (this._immutable || this.state !== CashlinkState.UNCHARGED)) {
+            throw new Error('Cashlink is immutable');
+        }
+        const messageBytes = Utf8Tools.stringToUtf8ByteArray(message);
+        if (!Nimiq.NumberUtils.isUint8(messageBytes.byteLength)) throw new Error('Message is too long');
+        this._messageBytes = messageBytes;
+    }
+
+    set networkClient(client: NetworkClient) {
+        this._networkClientResolver(client);
+    }
+
+    public static async create(): Promise<Cashlink> {
+        await loadNimiq();
+        const keyPair = Nimiq.KeyPair.derive(Nimiq.PrivateKey.generate());
+        return new Cashlink(keyPair, keyPair.publicKey.toAddress());
+    }
+
+    public static async parse(str: string): Promise<Cashlink | null> {
+        if (!str) return null;
+        try {
+            str = str.replace(/~/g, '').replace(/=*$/, (match) => new Array(match.length).fill('.').join(''));
+            const buf = Nimiq.BufferUtils.fromBase64Url(str);
+            await loadNimiq();
+            const keyPair = Nimiq.KeyPair.derive(Nimiq.PrivateKey.unserialize(buf));
+            const value = buf.readUint64();
+            let message;
+            if (buf.readPos === buf.byteLength) {
+                message = '';
+            } else {
+                const messageLength = buf.readUint8();
+                const messageBytes = buf.read(messageLength);
+                message = Utf8Tools.utf8ByteArrayToString(messageBytes);
+            }
+
+            return new Cashlink(
+                keyPair,
+                keyPair.publicKey.toAddress(),
+                value,
+                message,
+                CashlinkState.UNKNOWN);
+        } catch (e) {
+            console.error('Error parsing Cashlink:', e);
+            return null;
+        }
+    }
+
+    public static fromObject(object: CashlinkEntry): Cashlink {
+        return new Cashlink(
+            Nimiq.KeyPair.unserialize(new Nimiq.SerialBuffer(object.keyPair)),
+            Nimiq.Address.fromUserFriendlyAddress(object.address),
+            object.value,
+            object.message,
+            object.state,
+            object.date,
+            object.contactName,
+        );
+    }
+
+    private _network: Promise<NetworkClient>;
+    private _networkClientResolver!: (client: NetworkClient) => void;
```

Comment:
Ok, we can rename it `_getNetwork()`, but that does not remove the need for the external promise-resolver, because a promise can only be resolved from inside it's body, but the network is assigned to the Cashlink from an external action. So somewhere we need to store a reference to the resolver to then access it.

Reply:
Do you still want to rename it to `_getNetwork()`?

Reply:
Yes, `_getNetwork()` is definitely nicer in my opinion.
The need for having the `_networkClientResolver` could be removed by:
```ts
Object.defineProperty(this, 'networkClient', {
    set: (value) => resolve(value);
});
```
But you can also keep the `_networkClientResolver` if you prefer less magic.